app/src/main/java/com/github/multimatum_team/multimatum/MultimatumApp.kt

```diff
@@ -1,13 +1,109 @@
 package com.github.multimatum_team.multimatum
 
+import android.app.Activity
+import android.app.ActivityManager
 import android.app.Application
+import android.content.Context
+import android.content.SharedPreferences
+import android.os.Bundle
+import com.github.multimatum_team.multimatum.LogUtil.debugLog
+import com.github.multimatum_team.multimatum.LogUtil.logFunctionCall
+import com.github.multimatum_team.multimatum.MainSettingsActivity.Companion.PROCRASTINATION_FIGHTER_ENABLED_PREF_KEY
+import com.github.multimatum_team.multimatum.service.ProcrastinationDetectorService
 import dagger.hilt.android.HiltAndroidApp
+import javax.inject.Inject
 
 @HiltAndroidApp
-class MultimatumApp: Application() {
+class MultimatumApp : Application(), Application.ActivityLifecycleCallbacks {
 
     // This class is required by Hilt
-    // And code generation seems to be very sensitive to changes in the format of this file
-    // tl;dr: BLACK MAGIC, DO NOT TOUCH
+    // It is also useful to perform actions when the app is started
+
+    override fun onCreate() {
+        logFunctionCall()
+        super.onCreate()
+        registerActivityLifecycleCallbacks(this)
+    }
+
+    override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {
+        logFunctionCall("created ${activity.localClassName}")
+    }
+
+    override fun onActivityStarted(activity: Activity) {
+        logFunctionCall("started ${activity.localClassName}")
+    }
+
+    override fun onActivityResumed(activity: Activity) {
+        logFunctionCall("resumed ${activity.localClassName}")
+        val onForeground = isAppOnForeground()
+        debugLog(if (onForeground) "app is on foreground" else "app is on background")
+        if (onForeground){
+            stopProcrastinationDetectorIfActive()
+        }
+    }
+
+    override fun onActivityPaused(activity: Activity) {
+        logFunctionCall("paused ${activity.localClassName}")
+    }
+
+    override fun onActivityStopped(activity: Activity) {
+        logFunctionCall("stopped ${activity.localClassName}")
+        val onForeground = isAppOnForeground()
+        debugLog(if (onForeground) "app is on foreground" else "app is on background")
```

Comment:
Are the debugLog still useful ? If they are, you can ignore my comment

Reply:
I think the logs are not a problem for now, and may be useful when we debug the app in the following weeks. But yes at the very end of the project the ideal thing to do will probably be to remove them.

Reply:
We could probably setup the LogUtils so that the logs are only emitted on debug mode, and ignored in release mode.